### ISSUE TITLE
[TW-752] task ID no longer defaults to info

### ISSUE THREAD

**Issue by taskwarrior** [GothenburgBitFactory/taskwarrior#787]
_Christopher Roberts on 2011-09-19T11:16:22Z says:_ 

 (at)task [ID]@ no longer default to 'info', but instead returns a list of the one matching item.  There may be logical reasons why this has been changed, but it is unlikely to be what the user actually wants.

<pre>
$ task 100
[task next 100]

ID  Project Pri Due      A Age Urgency Description                                            
100         H   2011-9-2   2wk    19.6 Example task

1 task
</pre>

I would suggest that an exception is made when a single ID is entered (i.e. not a range), so that the default command is *not* used, and instead "info" is run.

This has the additional benefit of being what users have become used to in previous versions.

**Comment by taskwarrior** [comment]
Migrated metadata:
```
Created: 2011-09-19T11:16:22Z
Modified: 2014-05-31T14:21:46Z
```

**Comment by taskwarrior** [comment]
_Aikido Guy on 2011-09-19T13:51:16Z says:_ 

 For me, as a user, I would find it difficult to remember the exceptions; especially this one. For me, I can easily remember that the default command will be used if no command is specified. I like that behaviour... I can understand the request for backwards compatibility... but for new (future) users, they wouldn't necessarily understand the exception.

I'm in favour of forcing
<pre>task 100 info</pre>
because it is more consistent with other command usage and will make the command line state explicitly what action should be performed by TaskWarrior. That is, if there is no command, then the default command is used. This rule is simple. Why should 100 be supported but not, for example, the following:
<pre>task 100,101</pre>

What if a user has a default command that they want to use with 100? For example, in the case above, task has provided information about task 100 based on the default command. This seems consistent to me. If the user wants information about task 100, then they could state it explicitly.

Just my 2 cents :)

**Comment by taskwarrior** [comment]
_Aikido Guy on 2011-09-19T13:55:31Z says:_ 

 Sorry... I should have mentioned that if a particular user wants 'info' as the default, then they could change their default command to use 'info' and then if no command is issued 'info' would be used! :)

**Comment by taskwarrior** [comment]
_Christopher Roberts on 2011-09-19T14:10:20Z says:_ 

 Maybe I have been corrupted by the old usage, but it feels completely logical to me that when you call task on a single ID that it displays the content of that single ID.  I can't see any benefit to listing a single item.

I wouldn't wish to change my default command, as it is only this one isolated example where "info" makes sense as the default.

I think this is one of those situations where being consistent actually is unintuitive - you naturally expect something different when you call a command on a single ID than on a range of IDs.

In short, if you called task for a range of IDs - you would always want a list response; whereas if you called task for a single ID you would always want the info response - why else would you be calling task on a single ID?

But that's my view - feel free to close the issue if you are still not in agreement - I'm sure I can learn to live with the new functionality!!

**Comment by taskwarrior** [comment]
_Aikido Guy on 2011-09-19T14:22:05Z says:_ 

 I'm not a developer! Just a user with a point of view!! :) I just wanted the developers to know my point of view too! :)

**Comment by taskwarrior** [comment]
_Eric Fluger on 2011-09-19T17:36:18Z says:_ 

 No strong feelings about this myself.  Just wondering if it might make sense for task when given an ID number alone to default to printing a message like: 

<pre>
   No command provided.  No default command set.
   To set a default command please < do thus and so, etc >. 
   To suppress this message and have task do nothing in the absence of a command please < do something else >.
</pre>

So each user is encouraged to set their own value for the default command.

**Comment by taskwarrior** [comment]
_Christopher Roberts on 2011-09-19T19:24:46Z says:_ 

 Aikido Guy wrote:
> I'm not a developer! Just a user with a point of view!! :) I just wanted the developers to know my point of view too! :)

Ah apologies, I jumped to conclusions!

Eric Fluger wrote:
> No strong feelings about this myself.  Just wondering if it might make sense for task when given an ID number alone to default to printing a message like: 
> 
> [...]
> 
> So each user is encouraged to set their own value for the default command.

I really don't think changing the default command is helpful.  The only logical default command for multiple items is 'list', and the only logical default command for single items is 'info'.  Not sure what there is to be gained giving users a choice, beyond the ability to edit the configuration if they really want.  But maybe I'm missing something and there are good alternative commands - I've only been using task for a few weeks - but that's how it seems to me.

**Comment by taskwarrior** [comment]
_Aikido Guy on 2011-09-19T19:34:48Z says:_ 

 Sounds to me like you want something like:
<pre>default.command.single=info
default.command.multiple=next</pre>

**Comment by taskwarrior** [comment]
_Christopher Roberts on 2011-09-19T19:48:07Z says:_ 

 I like that idea :)

**Comment by taskwarrior** [comment]
_Aikido Guy on 2011-09-19T20:33:05Z says:_ 

 hhmm... what if :)

The following line does not have a command, so the default command will be used:
<pre>>task project:Home</pre>

ah, but which default command should we use in this case? food for thought... :)

suggestions anyone?

**Comment by taskwarrior** [comment]
_David Patrick on 2011-09-20T01:00:55Z says:_ 

 Aikido Guy wrote:
> suggestions anyone?

To me, this one's too easy. 
The user gets to define a default command, in case none is supplied. This is usually a listing, maybe custom, but some form of listing report UNLESS the only argument supplied is an ID, or a set of IDs, in which case (i.m.h.o.) the "info" command should be used instead. The logic being that if only an ID is supplied (or even in the case of several IDs) the listing report makes little sense, like a list of one (or a few) and if that's what you mean (show me a list of this one task) then "task ID list" should still do that. 

what's at stake is the very speedy details available using "t !D" 

you asked ;-)

**Comment by taskwarrior** [comment]
_Paul Beckingham on 2011-09-20T01:48:36Z says:_ 

 My opinion is this:  Consistency is good.  Special cases are bad.  I want to increase the former and decrease the latter.  2.0 is a huge step in that direction.

Prior to 2.0, ID was the only means of addressing tasks.  The enhancement (thanks to John Florian) of interpreting "task <id>" as "task <id> info" not only made sense, but was a genuine time-saver.

Things are different now.  ID is not the only way of addressing a task.  We have UUID too.  Not only that, but every write command supports filters that specify tasks, and therefore is capable of bulk operations.  That means that a single ID is only one of dozens of ways to specify tasks.  Should every filter that resolves to a _single_ task assume the 'info' command?  How about every filter that resolves to two tasks?

I used the "task <id>" form a lot, and liked it, but I don't see how to support it in a consistent manner.

**Comment by taskwarrior** [comment]
_Christopher Roberts on 2011-09-20T07:59:23Z says:_ 

 Well that is a shame, but at least I can understand the reasons now.  I am sure in time I will type 'info' without a second thought!

**Comment by taskwarrior** [comment]
_Christopher Roberts on 2011-09-20T08:00:56Z says:_ 

 I can't see how to close this issue, perhaps an admin needs to do that?

**Comment by taskwarrior** [comment]
_Paul Beckingham on 2011-09-20T12:43:54Z says:_ 

 I would like to keep this open, because I'm hopeful we can find a way to make it work consistently.  We're just not there yet.

**Comment by taskwarrior** [comment]
_Nikola Petrov on 2012-01-09T07:14:19Z says:_ 

 Why don't you write a bash wrapper that checks if the arguments supplied is just one and is a number? That seems reasonable solution and will work for all those people that want the old behaviour.

**Comment by taskwarrior** [comment]
_Nikola Petrov on 2012-01-09T10:50:56Z says:_ 

 Something like this should be enough if put in your rc file. Not sure if this will work zsh users though:
@
task(){
    if [[ $# -eq '1' ]] && ! [[ "$1" =~ ^[0-9]+$ ]] ; then
        command task $1 info
    else
        command task $@
    fi
}
@

**Comment by taskwarrior** [comment]
_Nikola Petrov on 2012-01-09T10:54:43Z says:_ 

 It seems i am not able to edit my comment nor i can provide the code here. Here is a gist instead - https://gist.github.com/1582477

**Comment by taskwarrior** [comment]
_Paul Beckingham on 2012-01-09T12:42:59Z says:_ 

 Nikola Petrov wrote:
> Why don't you write a bash wrapper that checks if the arguments supplied is just one and is a number?

That ignores the default command.

**Comment by taskwarrior** [comment]
_Paul Beckingham on 2012-01-28T22:12:51Z says:_ 

 Rejecting because this is complex, and I don't feel we have a good solution.

**Comment by taskwarrior** [comment]
_Paul Beckingham on 2014-05-31T14:19:43Z says:_ 

 With 2.4.0 and the new command line parser, this is back on the table.

**Comment by taskwarrior** [comment]
_Paul Beckingham on 2014-05-31T14:21:46Z says:_ 

 Finally.
